File: EduSphere_backend/enrollment-service/src/main/java/com/edusphere/enrollment/service/EnrollmentService.java
package com.edusphere.enrollment.service;

import com.edusphere.enrollment.dto.request.EnrollRequest;
import com.edusphere.enrollment.dto.response.EnrollmentResponse;

import java.util.List;
import java.util.UUID;

public interface EnrollmentService {

    EnrollmentResponse enroll(EnrollRequest request);

    EnrollmentResponse selfEnroll(UUID userId, String userRole, UUID courseId);

    EnrollmentResponse requestEnrollment(UUID studentId, UUID courseId);

    List<EnrollmentResponse> getPendingRequests();

    List<EnrollmentResponse> getStudentPendingRequests(UUID studentId);

    EnrollmentResponse approveEnrollment(UUID enrollmentId);

    void rejectEnrollment(UUID enrollmentId);

    List<EnrollmentResponse> getEnrollmentsByUser(UUID userId);

    List<EnrollmentResponse> getEnrollmentsByCourse(UUID courseId);

    void unenroll(UUID enrollmentId);

    List<EnrollmentResponse> getStudentEnrollments(UUID studentId);

    List<EnrollmentResponse> getInstructorEnrollments(UUID instructorId);

    boolean isEnrolled(UUID userId, UUID courseId);

    long getTotalEnrollmentCount();
}

File: EduSphere_frontend/src/services/enrollmentService.js
import api from './api'

export const enrollmentService = {
  // ADMIN/COORDINATOR: manual enroll
  enroll: (data) => api.post('/enrollments', data),

  // STUDENT/INSTRUCTOR: self-enroll — courseId is a query param not body
  selfEnroll: (courseId) => api.post('/enrollments/self', null, { params: { courseId } }),

  // STUDENT: request enrollment (requires coordinator approval)
  requestEnrollment: (courseId) => api.post('/enrollments/request', null, { params: { courseId } }),

  // COORDINATOR/ADMIN: get pending enrollment requests
  getPendingRequests: () => api.get('/enrollments/pending'),

  // COORDINATOR/ADMIN: approve enrollment request
  approveEnrollment: (enrollmentId) => api.post(`/enrollments/${enrollmentId}/approve`),

  // COORDINATOR/ADMIN: reject enrollment request
  rejectEnrollment: (enrollmentId) => api.post(`/enrollments/${enrollmentId}/reject`),

  // ADMIN/COORDINATOR/INSTRUCTOR: get enrollments for a course
  getEnrollmentsByCourse: (courseId) => api.get('/enrollments', { params: { courseId } }),

  // Get enrollments for a specific student
  getStudentEnrollments: (studentId) => api.get(`/students/${studentId}/enrollments`),

  // Get pending enrollment requests for a student
  getStudentPendingRequests: (studentId) => api.get(`/students/${studentId}/enrollments/pending`),

  // Get enrollments for a specific instructor
  getInstructorEnrollments: (instructorId) => api.get(`/instructors/${instructorId}/enrollments`),

  // Check enrollment status (internal use)
  checkEnrollment: (userId, courseId) =>
    api.get('/enrollments/check', { params: { userId, courseId }, silentError: true }),

  // ADMIN: unenroll
  unenroll: (enrollmentId) => api.delete(`/enrollments/${enrollmentId}`),

  // ADMIN: get past (soft-deleted) enrollments
  getPastEnrollments: () => api.get('/enrollments/past'),
}
